python/revng/tupletree.py

```diff
@@ -128,6 +128,18 @@ def __setattr__(self, key, value):
         super().__setattr__(key, value)
 
 
+class UpcastableStructBase(StructBase):
+    _children: List[Type] = []
+
+    @classmethod
+    def from_dict(cls, **kwargs):
+        if UpcastableStructBase in cls.__bases__ and "Kind" in kwargs:
```

Comment:
That wouldn't work
e.g. `FunctionEdgeBase(UpcastableStructBase)` -> `FunctionEdge(FunctionEdgeBase, StructBase)`
so `FunctionEdge` is a subclass of `UpcastableStructBase` indirectly

Reply:
I'm not sure why that would be a problem, actually, I'm not sure why we need the test at all.
Say we have `Base` and `Derived(Base)`. `cls` is `Base` when the function is called like `Base.from_dict(...)`, as is the case during deserialization. Even if someone called `Derived.from_dict` it would still work (*), as Derived also inherits the `_children` from `Base`:
```
class A:
  pass
class B:
  pass
A.x = 1
assert B.x == 1
```

Granted, this would mean that `Derived.from_dict` could create an instance of something else which admittedly might be a bit unexpected, is this what you mean?

Because of this reasoning I argue we actually might want to raise an exception if `Kind` is not in kwargs, I don't see how that could not be an error, but I guess that the base class `from_dict` fill eventually do that anyway.

Reply:
The problem is that without `__bases__` there will be a loop.
Example:

1. `from_dict` arrives at a `FunctionEdgeBase` and calls its `from_dict`
2. in `FunctionEdgeBase.from_dict` we have `UpcastableStructBase` in `__bases__` and a `Kind`, we call the relative class, let's say `FunctionEdge`
3. now we are in `FunctionEdge.from_dict` and we want to bail back in `StructBase.from_dict`, if we check `issubclass` it will return `True` since we have the chain `FunctionEdge -> FunctionEdgeBase -> UpcastableStructBase` and we go back to (2) whereas `__bases__` will be `[ FunctionEdgeBase, StructBase ]` and so the check passes